usersession/userd/launcher.go

```diff
@@ -170,6 +185,216 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
+// DBus interface. The desktopFileID is described here:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+func (s *Launcher) OpenDesktopEntryEnv(desktopFileID string, env []string, sender dbus.Sender) *dbus.Error {
+	desktopFile, err := desktopFileIDToFilename(existsOnFileSystem, desktopFileID)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	err = verifyDesktopFileLocation(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	exec_command, err := readExecCommandFromDesktopFile(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	args, err := parseExecCommand(exec_command)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		if !strutil.ListContains(allowedEnvVars, strings.SplitN(e, "=", 2)[0]) {
+			return dbus.MakeFailedError(fmt.Errorf("Supplied environment variable %q is not allowed", e))
+		}
+
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	// XXX: this avoids defunct processes but causes userd to persist
+	// until all children are gone (currently, this is not a problem since
+	// userd is long running once started)
+	go cmd.Wait()
+
+	if cmd.Run() != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
+	}
+
+	return nil
+}
+
+type fileExists func(string) bool
+
+func existsOnFileSystem(desktopFile string) bool {
+	fileStat, err := os.Stat(desktopFile)
+
+	// We only support files in /var/lib/snapd/desktop/applications and they should
+	// all be regular files.
+	return err == nil && fileStat.Mode().IsRegular()
+}
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+// Per https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id,
+// if desktop entries have dashes in the name ('-'), this could be an indication of subdirectories, so search
+// for those too. Eg, given foo-bar_baz_norf.desktop the following are searched for:
+//   o .../foo-bar_baz-norf.desktop
+//   o .../foo/bar_baz-norf.desktop
+//   o .../foo/bar_baz/norf.desktop
+//   o .../foo-bar_baz/norf.desktop
+// We're not required to diagnose multiple files matching the desktop file ID.
+func findDesktopFile(desktopFile_exists fileExists, base_dir string, splitFileId []string) (string, error) {
+	desktopFile := filepath.Join(base_dir, strings.Join(splitFileId, "-"))
+
+	if desktopFile_exists(desktopFile) {
+		return desktopFile, nil
+	}
+
+	// Iterate through the potential subdirectories formed by the first i elements of the desktop file ID.
+	// Maybe this is overkill: At the time of writing, the only use is in desktopFileIDToFilename() and there
+	// we're only checking dirs.SnapDesktopFilesDir (and not all entries in $XDG_DATA_DIRS).
+	// For dirs.SnapDesktopFilesDir snapd will not create any subdirectories.
```

Comment:
This is an incomplete sentence.

Reply:
Where is the incomplete sentence?

Reply:
> Where is the incomplete sentence?

"// For dirs.SnapDesktopFilesDir snapd will not create any subdirectories."

Reply:
I'm still not seeing what the problem is: Why is that incomplete?

Reply:
Maybe there is missing punctuation? Maybe this is better reworded as:
```
// we're only checking dirs.SnapDesktopFilesDir (and not all entries in $XDG_DATA_DIRS)
// and snapd will not create any subdirectories in dirs.SnapDesktopFilesDir.
```